### PR TITLE
Fix ADS observer behaviour on Fitting tab in Engineering Diffraction UI

### LINKED ISSUE #32837: Fix ADS behaviour for bgsub workspaces fitting tab EngDiff UI
Various issue relating to changes introduced when switched to keeping track of bgsub workspace separately for project save to work. 
At the minute the remove and replace don't work for bgsub workspaces and the `data_presenter` does a lot of the heavy lifting. I think we need to handle e.g.replacements in the model and keep track of both workspaces with custom class - here is a suggestion (no warranty!).

```
class WorkspaceRecord:
    def __init__(self, ws_name, bgsub_ws=None, bg_params=None):
        self.name = ws_name
        self.bgsub_ws = bgsub_ws
        self.bg_params = bg_params  # [isSub, niter, xwindow, doSG]

    def get_active_ws(self):
        if self.bgsub_ws and self.bg_params[0]:
            # first element is isSub checkbox
            return self.bgsub_ws
        else:
            return self.name


class WorkspaceRecordContainer:
    def __init__(self):
        self.dict = {}
        
    def __getitem__(self, key):
        return self.dict[key].name

    def add(self, ws_name, *args, **kwargs):
        self.dict[ws_name] = WorkspaceRecord(ws_name, *args, **kwargs)

    def get_loaded_workpace_names(self):
        return list(self.dict.keys())

    def get_bgsub_workpace_names(self):
        return [w.bgsub_ws for w in self.ws.values()]

    def is_workspace_loaded(self, ws_name):
        return ws_name in self.get_loaded_workpace_names()

    def is_workspace_bgsub(self, ws_name):
        return ws_name in self.get_bgsub_workpace_names()

    def is_workspace_contained(self, ws_name):
        return self.is_workspace_loaded(ws_name) or self.is_workspace_bgsub(ws_name)

    def get_bgsub_ws(self, ws_name):
        ws_loaded = self._get_loaded_ws_name(ws_name)
        if ws_loaded:
            return self.dict[ws_loaded].get_bgsub_ws()

    def get_active_ws(self, ws_name):
        ws_loaded = self._get_loaded_ws_name(ws_name)
        if ws_loaded:
            return self.dict[ws_loaded].get_active_ws()

    def _get_loaded_ws_name(self, ws_name):
        if self.is_workspace_loaded(ws_name):
            return ws_name
        elif self.is_workspace_bgsub(ws_name):
            return next((key for key, val in self.dict.items() if val.bgsub_ws == ws_name), None)

    def rename(self, old_ws, new_ws):
        ws_loaded = self._get_loaded_ws_name(old_ws)
        if ws_loaded:
            if ws_loaded == old_ws:
                self.dict[ws_loaded].name = new_ws
            else:
                self.dict[ws_loaded].bgsub_ws = new_ws

    def pop(self, ws_name):
        ws_loaded = self._get_loaded_ws_name(ws_name)
        return self.dict.pop(ws_loaded)

    def set_bg_params(self, ws_name, bg_params):
        ws_loaded = self._get_loaded_ws_name(ws_name)
        if ws_loaded:
            self.dict[ws_loaded].bg_params = bg_params

    def set_bgsub_ws(self, ws_name, ws_bgsub):
        ws_loaded = self._get_loaded_ws_name(ws_name)
        if ws_loaded:
            self.dict[ws_loaded].self.bgsub_ws = ws_bgsub
```

### PR DESCRIPTION
**Pull request by DannyHindson**
**Description of work.**

The fitting tab creates ~3 workspaces when you load a workspace:
a) the focussed workspace the user has loaded
b) a derived workspace that has background subtracted
c) a log group workspace

If the user renamed or deleted any of these the UI would often crash or log a nasty looking error message. The ADS observer behaviour has been reworked to sort this out. This relies on a new class called WorkspaceContainer in the fitting tab's model that links together the a) and b) workspaces in the list above

As part of this I noticed that the serial and sequential fit weren't taking into account the Subtract BG checkbox which indicates whether a) or b) is active for each loaded workspace. So I've fixed this bug also.

**To test:**

Open up the Engineering Diffraction interface
Go to the Fitting Tab
Load a focused data file - you can load the one in the attached zip file
[ENGINX_299080_307521_bank_2_TOF.zip](https://github.com/mantidproject/mantid/files/7556728/ENGINX_299080_307521_bank_2_TOF.zip)
Check the "Plot" and "Subtract BG" checkbox on the row created for this file in the table
Try renaming or deleting the workspaces in ADS and check the UI does something sensible

Fixes #32837.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: qt/python/mantidqtinterfaces/mantidqtinterfaces/Engineering/gui/engineering_diffraction/tabs/common/calibration_info.py [hunk 1]
```diff
@@ -127,7 +127,12 @@ def load_relevant_calibration_files(self, output_prefix="engggui"):
         """
         filepath = path.splitext(self.prm_filepath)[0] + '.nxs'  # change extension to .nxs
         self.calibration_table = output_prefix + "_calibration_" + self.get_group_suffix()
-        Load(Filename=filepath, OutputWorkspace=self.calibration_table)
+
+        try:
+            Load(Filename=filepath, OutputWorkspace=self.calibration_table)
+        except Exception as e:
+            logger.error("Unable to loading calibration file " + filepath + ". Error: " + str(e))
```
Comment (DanielMurphy22) [inline qt/python/mantidqtinterfaces/mantidqtinterfaces/Engineering/gui/engineering_diffraction/tabs/common/calibration_info.py:134]:
```suggestion
            logger.error("Unable to load calibration file " + filepath + ". Error: " + str(e))
```
Sorry to nitpick

File: qt/python/mantidqtinterfaces/mantidqtinterfaces/Engineering/gui/engineering_diffraction/engineering_diffraction_io.py [hunk 2]
```diff
@@ -35,11 +35,10 @@ def encode(self, obj, _=None):  # what obj = EngineeringDiffractionGui
             obj_dic["settings_dict"] = presenter.settings_presenter.settings
         else:
             obj_dic["settings_dict"] = presenter.settings_presenter.model.get_settings_dict(SETTINGS_KEYS_TYPES)
-        if data_widget.presenter.get_loaded_workspaces():
-            obj_dic["data_loaded_workspaces"] = [*data_widget.presenter.get_loaded_workspaces().keys()]
+        if data_widget.model._data_workspaces.get_ws_names_dict():
+            obj_dic["data_workspaces"] = data_widget.model._data_workspaces.get_ws_names_dict()
             obj_dic["fit_results"] = data_widget.model.get_fit_results()
             obj_dic["plotted_workspaces"] = [*data_widget.presenter.plotted]
-            obj_dic["background_params"] = data_widget.model.get_bg_params()
```
Comment (DanielMurphy22) [inline qt/python/mantidqtinterfaces/mantidqtinterfaces/Engineering/gui/engineering_diffraction/engineering_diffraction_io.py:42]: Why is it fine to remove this background_params entry, also on line 72 of this file?

**Review by DanielMurphy22 — CHANGES_REQUESTED** [review]
My manual test of the interface went really well, renaming and deleting works as expected. My code review only produced a question and an annoying nitpick (sorry!)

File: qt/python/mantidqtinterfaces/mantidqtinterfaces/Engineering/gui/engineering_diffraction/tabs/common/calibration_info.py (hunk 1 above)
Comment (DannyHindson) [reply qt/python/mantidqtinterfaces/mantidqtinterfaces/Engineering/gui/engineering_diffraction/tabs/common/calibration_info.py:134]: No problem!

**Review by DanielMurphy22 — APPROVED** [review]
Works really well! Thanks for your answer!

**Pull request merged**